src/components/GraphiQL.js

```diff
@@ -649,10 +654,16 @@ export class GraphiQL extends React.Component {
   }
 
   handleEditQuery = debounce(100, value => {
+    let operationName = this.state.operationName;
     if (this.state.schema) {
-      this._updateQueryFacts(value);
+      const newOperationName = this._updateQueryFacts(value);
+      // If `operationName` is updated as a result from updating query facts,
+      // treat it as a source of truth and update here accordingly.
+      if (newOperationName && newOperationName !== operationName) {
```

Comment:
This is repeating the logic on lines 684 below - that seems like this could introduce *more* possibilities for race conditions or split logic